protocol/binary/stream_reader.go

```diff
@@ -0,0 +1,445 @@
+		var buf bytes.Buffer
+		_, err := io.CopyN(&buf, sr.reader, int64(length))
+		if err == io.EOF {
+			// All EOFs are unexpected when streaming
+			err = io.ErrUnexpectedEOF
+		}
+
+		return buf.Bytes(), err
+	}
+
+	bs := make([]byte, length)
+	_, err = sr.read(bs)
+	return bs, err
+}
+
+// ReadStructBegin reads the "beginning" of a Thrift encoded struct.  Since
+// there is no encoding for the beginning of a struct, this is a noop.
+func (sr *StreamReader) ReadStructBegin() error {
+	return nil
+}
+
+// ReadStructEnd reads the "end" of a Thrift encoded struct.  Since
+// `ReadFieldBegin` will already be interpreting field-type of whether it's a
+// stop field or not, there is no real representation of a struct's end, making
+// this a noop.
+func (sr *StreamReader) ReadStructEnd() error {
+	return nil
+}
+
+// ReadFieldBegin reads off a Thrift encoded field-header returning that and a
+// 'bool' representing whether or not a field-value follows.
+// A 'false' without any error means that it has reached the stop-field.  There
+// is no guarantee that the field-header is valid in this case.
+func (sr *StreamReader) ReadFieldBegin() (stream.FieldHeader, bool, error) {
+	fh := stream.FieldHeader{}
+
+	fieldType, err := sr.ReadInt8()
+	if err != nil {
+		return fh, false, err
+	}
+
+	// fieldType 0 is the equivalent of the stop field, signaling the end
+	if fieldType == 0 {
+		return fh, false, nil
+	}
+
+	fieldID, err := sr.ReadInt16()
+	if err != nil {
+		return fh, false, err
+	}
+
+	fh.ID = fieldID
+	fh.Type = wire.Type(fieldType)
+	return fh, true, nil
+}
+
+// ReadFieldEnd reads the "end" of a Thrift encoded field  Since there is no
+// encoding for the end of a field, this is a noop.
+func (sr *StreamReader) ReadFieldEnd() error {
+	return nil
+}
+
+// ReadListBegin reads off the list header of a Thrift encoded list.
+func (sr *StreamReader) ReadListBegin() (stream.ListHeader, error) {
+	lh := stream.ListHeader{}
+
+	elemType, listSize, err := sr.readTypeSizeHeader()
+	if err != nil {
+		return lh, err
+	}
+
+	lh.Type = wire.Type(elemType)
+	lh.Length = int(listSize)
+	return lh, nil
+}
+
+// ReadListEnd reads the "end" of a Thrift encoded list.  Since there is no
+// encoding for the end of a list, this is a noop.
+func (sr *StreamReader) ReadListEnd() error {
+	return nil
+}
+
+// ReadSetBegin reads off the set header of a Thrift encoded set.
+func (sr *StreamReader) ReadSetBegin() (stream.SetHeader, error) {
+	sh := stream.SetHeader{}
+
+	elemType, setSize, err := sr.readTypeSizeHeader()
+	if err != nil {
+		return sh, err
+	}
+
+	sh.Type = elemType
+	sh.Length = setSize
+	return sh, nil
+}
+
+// ReadSetEnd reads the "end" of a Thrift encoded list.  Since there is no
+// encoding for the end of a set, this is a noop.
+func (sr *StreamReader) ReadSetEnd() error {
+	return nil
+}
+
+func (sr *StreamReader) readTypeSizeHeader() (wire.Type, int, error) {
+	elemType, err := sr.ReadInt8()
+	if err != nil {
+		return wire.Type(0), 0, err
```

Comment:
. Constant literals are untyped, so just 0 should work as a wire.Type here.

```suggestion
		return 0, 0, err
```